cmd/jumphost/create_test.go

```diff
@@ -0,0 +1,327 @@
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+type mockJumphostAWSClient struct {
+	jumphostAWSClient
+	mock.Mock
+}
+
+func (m *mockJumphostAWSClient) DescribeImages(ctx context.Context, params *ec2.DescribeImagesInput, optFns ...func(options *ec2.Options)) (*ec2.DescribeImagesOutput, error) {
+	args := m.Called(ctx, params)
+	return args.Get(0).(*ec2.DescribeImagesOutput), args.Error(1)
+}
+
+func (m *mockJumphostAWSClient) AuthorizeSecurityGroupIngress(ctx context.Context, params *ec2.AuthorizeSecurityGroupIngressInput, optFns ...func(options *ec2.Options)) (*ec2.AuthorizeSecurityGroupIngressOutput, error) {
+	args := m.Called(ctx, params)
+	return args.Get(0).(*ec2.AuthorizeSecurityGroupIngressOutput), args.Error(1)
+}
+
+func (m *mockJumphostAWSClient) DescribeSubnets(ctx context.Context, params *ec2.DescribeSubnetsInput, optFns ...func(options *ec2.Options)) (*ec2.DescribeSubnetsOutput, error) {
+	args := m.Called(ctx, params)
+	return args.Get(0).(*ec2.DescribeSubnetsOutput), args.Error(1)
+}
+
+func (m *mockJumphostAWSClient) CreateKeyPair(ctx context.Context, input *ec2.CreateKeyPairInput, opts ...func(*ec2.Options)) (*ec2.CreateKeyPairOutput, error) {
+	args := m.Called(ctx, input)
+	return args.Get(0).(*ec2.CreateKeyPairOutput), args.Error(1)
+}
+
+func TestFindLatestJumphostAMI(t *testing.T) {
+	mockClient := new(mockJumphostAWSClient)
+	jumphost := &jumphostConfig{awsClient: mockClient}
+	ctx := context.TODO()
+	mockImages := &ec2.DescribeImagesOutput{
+		Images: []types.Image{
+			{
+				ImageId:      aws.String("testImageID"),
+				CreationDate: aws.String("2023-04-01T12:00:00Z"),
+			},
+		},
+	}
+	tests := []struct {
+		name          string
+		mockError     error
+		expectedAMI   string
+		expectError   bool
+		errorContains string
+	}{
+		{
+			name:        "success_case_find_latest_AMI",
+			mockError:   nil,
+			expectedAMI: "testImageID",
+			expectError: false,
+		},
+		{
+			name:          "error_case_describeimages_fails",
+			mockError:     fmt.Errorf("errorDescribeImages"),
+			expectedAMI:   "",
+			expectError:   true,
+			errorContains: "failed to describe images in order to launch an EC2 jumphost: errorDescribeImages",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mockClient.On("DescribeImages", ctx, mock.Anything).
+				Return(mockImages, tt.mockError).Once()
+			ami, err := jumphost.findLatestJumphostAMI(ctx)
+			if tt.expectError {
+				assert.Error(t, err)
+				assert.Equal(t, tt.expectedAMI, ami)
+				assert.EqualError(t, err, tt.errorContains)
+			} else {
+				assert.NoError(t, err)
+				assert.Equal(t, tt.expectedAMI, ami)
+			}
+
+			mockClient.AssertExpectations(t)
+		})
+	}
+}
+
+func TestAllowJumphostSshFromIp(t *testing.T) {
```

Comment:
This test is sensitive for internet connection, due to the use of ip, err := determinePublicIp(). However due to the nature of the code, it work well.
So I have question for other reviewers. Is exiting without error when not able to connect to internet ok?